### ISSUE TITLE
Research and spells

### ISSUE THREAD

**Issue by hwoarangmy** [OpenDungeons/OpenDungeons#411]
I've just had a thought concerning spells. In #34, we said that some creatures could drop spells (which is, IMHO, a good thing).
Then, we could imagine that when we do a research in the library, when a spell is found, a spell item is created and the spell cannot be used. It has to be carried by a kobold to the dungeon temple to become available.
We could also imagine that when you kill a dungeon temple, the spells could be released to allow stealing them ^^

WDYT ?

Edit: See also: http://forum.freegamedev.net/viewtopic.php?f=37&t=6079

**Comment by Danimal696** [comment]
Safer than the library, not a bad idea. But the spell should be researcheable even if dropped, we can think of it like you have been lucky and skip that research or you got a spell you wont be able to research because of your campaign advancement

**Comment by Bertram25** [comment]
> Then, we could imagine that when we do a research in the library, when a spell is found, a spell item is created and the spell cannot be used. It has to be carried by a kobold to the dungeon temple to become available.

Yeah, I like the concept. :) Plus, it is aligning the library way to work with the forge.

> We could also imagine that when you kill a dungeon temple, the spells could be released to allow stealing them ^^

It's also an appealing idea, IMHO. This way, you have a direct bonus when destroying an opponent. :)

What do you think? @akien-mga @oyvindln @tomluchowski

**Comment by Danimal696** [comment]
Me like

**Comment by hwoarangmy** [comment]
> Yeah, I like the concept. :) Plus, it is aligning the library way to work with the forge.

Yes. Of course, that's the same if we choose to use the library only (ie if an enemy drops a spell, we will need to bring it to the library). But I guess most of the times, we will need to research to have a new spell and, thus, we won't see our kobolds carry the spells.
The bad point is that if we do as I said, the library won't get full after a while and we won't need a big library...

**Comment by Bertram25** [comment]
> The bad point is that if we do as I said, the library won't get full after a while and we won't need a big library...

It's a choice to make, but AFAIK, I thought the size of the library was used only to set the research speed. I mean the more active spots you have, the more creature can research, the research points you get. An ever-growing library doesn't make sense if we follow this principle.
But you're right, we should be able to see what spells were researched and that's where a spell/skill tree window is needed, IMHO. Plus, there will be the spell tab listing the active/battle spells that can be triggered by your evil hands...

What do you think about it?

**Comment by hwoarangmy** [comment]
IMHO, 1 creature per active spot is too much for the library. A 3x3 library would be enough for 5 creatures which is too many. That's why we should allow 1 creature per central active spot only IMHO and add a bonus for wall active spots.
We might even consider having this behaviour for all rooms since, AFAIK, only the training room allows for more than that. And I think it decreases too much the need for such rooms.

**Comment by Bertram25** [comment]
> IMHO, 1 creature per active spot is too much for the library. A 3x3 library would be enough for 5 creatures which is too many. That's why we should allow 1 creature per central active spot only IMHO and add a bonus for wall active spots.

It's definitely the way the forge is working so it's fine by me. :)

> We might even consider having this behaviour for all rooms since, AFAIK, only the training room allows for more than that. And I think it decreases too much the need for such rooms.

Can't say. I don't think it will hurt. What do you guys think about it?

**Comment by Danimal696** [comment]
yes, i also think its a good solution, when will be seeing the summon kobold spell? :)
As for the research tree ill post something on the forums

**Comment by hwoarangmy** [comment]
> when will be seeing the summon kobold spell? :)

I'm currently working on traps but if you import the new kobold model, I will work on the spell. Deal ?

**Comment by Danimal696** [comment]
We shall see ;D

**Comment by Bertram25** [comment]
> I'm currently working on traps but if you import the new kobold model, I will work on the spell. Deal ?

How evilly cunning of you @hwoarangmy ;)

**Comment by Bertram25** [comment]
Moving to 0.5.0 since it seems it is the plan.

**Comment by hwoarangmy** [comment]
> yes, i also think its a good solution, when will be seeing the summon kobold spell? :)

The kobold spell is in. I'm looking forward the kobold replacement ;-)

**Comment by Danimal696** [comment]
ok, ok, he will be the first after im done with exams next week

**Comment by hwoarangmy** [comment]
> he will be the first after im done with exams next week

cool ^^

**Comment by Bertram25** [comment]
> ok, ok, he will be the first after im done with exams next week

Watch out, internet never forgets. ;P

**Comment by hwoarangmy** [comment]
Do we agree with this behaviour?
- Creatures can work on library central active spots only. Wall active spots give a bonus (faster research) per wall active spot
- Once a research is done, a carryable research box will be created in the library and the corresponding room/spell will NOT be useable.
- The research box will then be carried by a worker to some room (ATM the dungeon temple but it can be easily changed later) to be usable
- There will be several paths of research (the famous evil paths) that will bring different rooms/spells 

That leads us to some stuff we have to sort out before going forward on research:
- How do we choose what to research ? Do we choose each item (if so, will all the creatures in the library be dropped from the room when a research is done) or only a path ?
- If we agree that workers have to bring the research box to the dungeon temple before it gets useable, that means that it may never get there (if no worker for example) or that it may be stolen by another player. In this case, how do we do to stop the current research and work again on the lost one ?

My answers:
- We choose a path because I would find annoying to have to go set something on the library because a research ended during a battle or while I'm doing something else.
- I would set the seat to the library it popped when the research is done and, when a worker tries to carry it, I would set its seat to the research box. This way, as soon as an ennemy worker steals my research box, my library will know that it should start the same research again.

@Bertram25 @Danimal696 @akien-mga @eugeneloza @oyvindln WDYT ?

**Comment by eugeneloza** [comment]
I have several ideas.
1. (The stuff I want to implement in my Project Helena game) The research is made at the library. BUT to research you have to find a determined set of prototypes first. I.e. you can't research LVL7 engine until you kill 10 enemies with LVL7 engines which are extracted and analyzed. To say it more OD-specific way, to research something you have to find a set of artefacts. Most of the artefacts can be obtained by searching the dungeon and carried to the library. Alternative way is to kill creatures of enemy posessing the artefact/spell/technology which a) randomly drop an artefact or b) each yields a stable portion of the artefact (e.g. 1 of 10 thunder arrwows) or c) yield a lower-quality artefact, which requires 3x or 5x more time to research or d) only specific monster type provides the artefact (e.g. cobold spell = by cobolds summoned with that spell / or to steal the enchantment technology the player must slay 3 of creatures using the weapon enchanted that way) - making this way less efficient than a direct artefact find but still usable. When all the research artefacts are in place, the research is automatically started.
2. Especially in junction with (1): The technology is separated into two classes: common, specific and unique. Common technologies can be researched by any player at any time. Specific are different for each faction and so cannot be researched by other faction (other player), only stolen. And unique - follows Master of Orion 1 conept - The first to research the technology makes it impossible for other players to start the research (they still may finish the research if already started). I.e. the first to grab the research is the first to succeed. Such technology may be only stolen afterwards.

**Comment by eugeneloza** [comment]
Another idea is to make a room specificly for storage of the research results. It may be not hard on design (I think I can do that if needed) - just colums with glowing colord spheres each representing one research and requiring enough space to store them. That would increase the rooms diversity.

**Comment by hwoarangmy** [comment]
That's an interesting vision and we can use this kind of concept for campaign maps but for OD in general, I'm not sure it fits well. Games such as UFO AI uses this kind of stuff but we are talking of games that length 10 hours or more.
In OD, if we want to have an interesting multiplayer (which is the most important IMHO), we should aim at games from 15min to 1hour at most which is too short for this.
But the concept is interesting for a campaign (at each level, we find artefacts and that unlocks research for spells/rooms).

> Another idea is to make a room specificly for storage of the research results.

Anyway changing the destination room of the research boxes will be very easy so we don't have to bother too much about it. If we want to use a specific room or the dungeon temple, we can switch easily.

**Comment by eugeneloza** [comment]
Yes, I was thinking of campagin/singleplayer first of all :). It definitely does not fit multiplayer or small singleplayer maps :)

**Comment by hwoarangmy** [comment]
Yes, for campaign, we need to find a main story line (even if in a game such as OD, we don't need something too complex) and unlock rooms/items one after each other depending on what happens in the story (finding some artefacts, killing special monsters, ...) to make the campaign feel like a whole and not a pack of maps.

**Comment by Bertram25** [comment]
> - Creatures can work on library central active spots only. Wall active spots give a bonus (faster research) per wall active spot

Yes.

> - Once a research is done, a carryable research box will be created in the library and the corresponding room/spell will NOT be useable.
> - The research box will then be carried by a worker to some room (ATM the dungeon temple but it can be easily changed later) to be usable

Yes. I'd even say that the temple should be usable as a gift boxes room, and that using a special room to store those elsewhere could come as a choice in the abilities along the path. The new room would then become the new place to store spells and abilities and the workers would then move them once the room is created if there is enough place in it. But that's for later.

> - There will be several paths of research (the famous evil paths) that will bring different rooms/spells
> 
> Yes, for campaign, we need to find a main story line (even if in a game such as OD, we don't need something too complex) and unlock room/items one after each other depending on what happens in the story (finding some artefacts, killing special monsters, ...)

First of all, completely agree with the whole concept of the campaign where you have to follow objectives to unlock new abilities, the objectives would thus be linked to the abilities.
And to link the campaign philosophy with the single playthrough one, I would split the paths into 2 or 3 segments.
Then, the player could choose the path segment taken when starting the game until he reaches the next segment, make another choice, and so forth.
Each segment combination could bring interesting mix of gameplay while it would be upon us to set up relevant possibilities, of course.
Of course, special monsters, artifact, or even unique rooms to claim, could give special abilities. :)
Some fourth and fifth path segments with unique spells could be unlocked with special monsters, ... (thus more used for the later campaign maps, I guess), or as a game rule mod, or when placing special rooms/gift boxes, to increase replayability?

> I would set the seat to the library it popped when the research is done and, when a worker tries to carry it, I would set its seat to the research box. This way, as soon as an ennemy worker steals my research box, my library will know that it should start the same research again.

I find this good and would do the same for crafted traps if it isn't the case already.

WDYT?

**Comment by hwoarangmy** [comment]
> I find this good and would do the same for crafted traps if it isn't the case already.

To be honest, the idea of using the seat for reserach boxes popped while I was writting. This was an unanswered  question when I began to write ^^
But I thought the same about traps. It will be easier than using the room they are in ^^

EDIT: I created an issue #489

**Comment by Danimal696** [comment]
My two cents:
- The temple is a good place for storing, safest on the dungeon, will be pretty satisfying when you destroy one and all spell pop out ready for the taking
- look here:
  http://forum.freegamedev.net/viewtopic.php?f=37&t=6079&p=62436#p62436
  Basically :
  The colour represent focus (or Paths of Evil) Red: strength - minions buffing ,agressive, Grey: neutral - versatile and defensive , Blue: magic - lots of spells.
  The darker rows show research difficulty tiers, they could be locked on campaign until completing some level and awarding access to the next tier.
  First row is free and pre-unlocked, since they are the most basic and essential rooms, we should block the research button to the condition of having a library
  You need to research at least 1 item to proceed to a lower row of the same branch (Path)
  Overall you could research all items of the low and medium tier for versatility or rush to the end of the highest tier in one path/colour and get some heavy firepower if you are faster than your opponents, or research everything on the same level if you feel god-like.

So going by that, in each tier to study a topic is much more expensive than the previous, so specialization is a need if you dont have much time.

**Comment by hwoarangmy** [comment]
ok. That brings us to how to choose what to research ? Should we choose a path ? Or each spell ? IMHO, it would be better to be able to choose somehow several researches at once to avoid to have to open the research tree each time a research is completed. That could be done by choosing a path or by picking each wanted spell (of course, spells that cannot be researched because requirements are not met will only start when available).

**Comment by akien-mga** [comment]
> That could be done by choosing a path or by picking each wanted spell (of course, spells that cannot be researched because requirements are not met will only start when available).

I am thinking of two possibilities:
- Civilization-like: you select one research (potentially one that you can't research right now, but for which you need to climb up its path), and then the library will automatically research all research topics that lead to the one you selected. That could allow to research either a whole path if you click the last research, or part of it if you want to specialise in several paths.
- À la carte: you define all research topics that you want to study in a specific order. It could be managed in two ways:
  - Highlighting: each time you double click a research topic, it gets highlighted and added to the bottom of your to-research list. That might end up confusing if you select lots of research topics though, so maybe adding numbers that specify the research order should be done on the GUI side. If you double-click a selected item, it gets removed from the to-research list and the following topics are brought up.
  - Research list window: similar to the one above but you would actually see each topic in a specific GUI element, where you could reorder things as you want.

**Comment by eugeneloza** [comment]
Maybe, in order there would be no 'idle' research, the creatures research smth at random, inefficiently. Clicking a room brings up guided research dialogue.
Another possibility - bring up research selection screen at each creature drop to the library. As soon as the creature finishes assigned job, it walks away idle. Scholars would sometimes go and start the research at random.

**Comment by Danimal696** [comment]
i think being able to select what you want researched is for the best, i would vote for "À la carte"

**Comment by Bertram25** [comment]
If we go to the "A la carte", which is fine in it self, I wouldn't add a research list but visually show on the research tree what is being researched and what is pending.
At least in multiplayer, (I mean when not in the campaign where the next thing will be available through scenario only), I would either:
- (Preferred option) When the research queue is empty, the player should be warned about it, just like it is the case when you don't have any treasury. After all, it's the player role to watch out for necessary stuff.
  The player could still define/remove more than one research topic by clicking on several options. We could then add a digit next to the research icon to show the order.
- (Secondary option or beginner option) We could propose a default research complete path that the player could later change, and why not use the same logic for AI.

This way, new players would have a standard, easy while sane research queue and advanced players would have the freedom to choose whatever they want.

**Comment by hwoarangmy** [comment]
In the screen provided by @Danimal696, it is not clear what depends on what (or we may want to stick to 1 item has to be searched per line per path to be able to go to next line).

For the record, the image is:
![research tree](https://cloud.githubusercontent.com/assets/7493706/6152122/fa4595b8-b21a-11e4-8eb2-11b585c703b8.png)

**Comment by akien-mga** [comment]
Well as far as understand:

```
for x in range(0, 9):
    for y in range(1, 6):
        tile[x][y] depends on tile[x][y-1]
```

**Comment by Danimal696** [comment]
> In the screen provided by @Danimal696, it is not clear what depends on what (or we may want to stick to 1 item has to be searched per line per path to be able to go to next line).

Yes that was my idea, you need at least 1 in the previous step to go down another step, the darker areas would correspond to research tiers that could be unlocked in the campaign (or by map setting configuration?) Research icon would have a tooltip explaining what they are, there could be lines joining topics from different trees to allow the research of a given one (say, you need to research spike trap and poison spell to be able to research the poison trap) but those would be only few.

I also like the predefined research for newbies, the AI can use the same to become an arquetipe, it would be cool to recieve tauntings from a brute ai or a mage one in different voices and styles.

**Comment by Bertram25** [comment]
I guess we more or less agree on the overall idea of the research tree.
What @hwoarangmy is pointing out now, I guess, is that the research tree in its current form is a bit incomplete or lacking explanation. 
Could you describe every still unused icons for a start?

**Comment by Danimal696** [comment]
there are some explanations on the forum post

**Comment by Bertram25** [comment]
Then bridges and the new forge (for equipment) are missing.

**Comment by Danimal696** [comment]
and doors and outpost

**Comment by Danimal696** [comment]
it was to have a guide of what it could be, many things could be added to it yet. Lest just stick with the implemented ones and prepare the place for the future ones

**Comment by Bertram25** [comment]
k, anyway we can't add random useless button in the research tree so I couldn't agree more on the actual implementation.

**Comment by hwoarangmy** [comment]
@Bertram25 Do you want to go for the UI part ? If you want, I can do the client/server part. You are more used to CEGUI than I am. IMHO, the best is to think what we want to do from the UI and adapt the technical part to reach that.

**Comment by Danimal696** [comment]
to the ones who made the GUI, it would be needed to a have an icon for unresearched topics and a different one for researched ones; so its easy to know what has already been researched.

**Comment by akien-mga** [comment]
> to the ones who made the GUI, it would be needed to a have an icon for unresearched topics and a different one for researched ones; so its easy to know what has already been researched.

I think the simplest way to achieve that would be to grey out unresearched topics, the highlight them (with a shiny outlet or something) when selected for research, and display them properly lighted when researched.

**Comment by Bertram25** [comment]
@hwoarangmy Np, for a common effort. I'll handle the UI part with pleasure.
I simply need to know what we want. Or i go wild and do a complete personal first try based on the discussion. ;)

AFAIK, in the UI, we'll want to add/remove research in/from the queue. 
We'll want to see something telling us what is unavailable, available, being done and queued.
The available state of the, let's call that skill, will be the default state.
The unavailable state can be set by disabling the image button,
The being done and pending states should be done using some kind of icons.
EDIT: Ah, indeed, we'll need something when the skill is researched.

I'll look for them and show a mockup of what I have in mind, still today if I'm lucky.

**Comment by hwoarangmy** [comment]
I will add that there should be a message sent to the server when you have finished to set researches (after each clic if we want to validate each or after the whole if we want to check only once).
Then, the server will send a message back to the client with the researches that are (and will) be done.

Also, we have to keep in mind that the researches can be done (after the research box is in the dungeon temple) or undone (when the dungeon temple is killed or if we allow somehow to "eject" a research box).

**Comment by eugeneloza** [comment]
Let's also note, if we need a voice announcement(s) for 'research completed'. These may be generic 'Research completed' or specific 'Kobold spell has been researched'.

**Comment by Danimal696** [comment]
it would be better to have the generic for now, since we dont even know many of them we will have

**Comment by eugeneloza** [comment]
Ok! Added this one to the script.

**Comment by Bertram25** [comment]
Hey,

Just to say that I'm this, but I've got one notice to add:
To stay pragmatic and permit the implementation to happen asap, what about the fact that I add a very simple research tree window, so that you've got something you can code against @hwoarangmy ?
This way, I can add the fancy stuff afterward.
In the meantime, I'd like you to make the ready-to-research items in the default state, and the other ones disabled, simply.

WDYT?

**Comment by hwoarangmy** [comment]
> I add a very simple research tree window, so that you've got something you can code against @hwoarangmy

Good idea. I've also started to implement the research logic. If you could provide a version with only 1 or 2 spells and a stub in gui.cpp, it should be enough for me to test. And I will try to provide the needed so you have to bind the buttons you will had with the gui functions I will provide.

Anyway, I've thought a bit about research and I guess the tree (ie dependencies on researches) should be hardcoded since it will be linked to a static image (unless you tell me you want to go for a dynamic tree).
Moreover, do we want to have OR relationship between researches ? That would mean that we can open a new research by finishing 1 of 2 other researches.

**Comment by Bertram25** [comment]
> Anyway, I've thought a bit about research and I guess the tree (ie dependencies on researches) should be hardcoded since it will be linked to a static image (unless you tell me you want to go for a dynamic tree).

True. I don't think it's a good idea to start making this configurable, anyway. Ok, I'll deliver the code stub + layout asap.

> Moreover, do we want to have OR relationship between researches ? That would mean that we can open a new research by finishing 1 of 2 other researches.

Well, we'll have three columns, and N rows. When finishing a research on the the row X, it gives access to the row X+1 in the given column. At least, that's what I understood, there might be counter example once we've started playing with it, unfortunately.
WDYT?

**Comment by hwoarangmy** [comment]
> At least, that's what I understood

That's also what I understood from the research tree found by @Danimal696. But from the ones you shown (and we have choosen), it looked much like one (or several) researches unlocks new ones.
I guess the best would be to allow both AND and OR relationship to be able to adapt to both ^^

**Comment by Bertram25** [comment]
> I guess the best would be to allow both AND and OR relationship to be able to adapt to both ^^

True enough. :) The research layout + stub should be ready tomorrow night. I'll at least give you the simple layout by then.

**Comment by hwoarangmy** [comment]
@Bertram25 I've done the research logic. For now, it is hardcoded but ready for the gui. I will try to open a PR tonight and will comment how to use it ^^

**Comment by Bertram25** [comment]
As shown here: http://forum.freegamedev.net/viewtopic.php?p=62549#p62549

![research-tree-basic](https://cloud.githubusercontent.com/assets/765917/6252062/dcae03a0-b797-11e4-9502-66fa1ce0c650.png)

The layout file is also available there.

**Comment by hwoarangmy** [comment]
Nice. How do you display selected/already done researches ? And the order ?

**Comment by Danimal696** [comment]
i guess i need to look out for a book model

**Comment by Bertram25** [comment]
> How do you display selected/already done researches ?

I'll be honest, I don't actually know yet. I guess a check mark or something should appear on the bottom right of the researches done, and something like a cog should appear on pending ones. The unavailable ones should simply appear disabled, even if you should be able to click on them to add them to the pending queue.

For that, I'll add new ImagerySection in ImageButton or maybe I'll duplicate that one and make something like a ResearchButton look'n'feel class:
https://github.com/OpenDungeons/OpenDungeons/blob/development/gui/OpenDungeons.looknfeel#L4091

Then, I'll need to have a look at how I can change the layers used on the different button states here depending on the research state: https://github.com/OpenDungeons/OpenDungeons/blob/development/gui/OpenDungeons.looknfeel#L4130

In the meanwhile, only available researches will be enabled, the rest (What's done and unreachable) will be disabled until I've figured out how to dynamically change the layers in a button state.

EDIT: Useful link: http://static.cegui.org.uk/docs/0.8/classCEGUI_1_1StateImagery.html

**Comment by Danimal696** [comment]
a suggestion: 
topic available to be researched -> icon without borders, 
reseach finished -> silver border like screen, 
non available -> greyed topic

**Comment by Bertram25** [comment]
@Danimal696 I'd still use more meaningful icon on the widget to show the state. Showing borders will definitely give a hint but while I'm at it, let's make it good. :)
